aptos-move/aptos-vm-types/src/change_set.rs

```diff
@@ -219,104 +406,106 @@ impl VMChangeSet {
+                "Cannot convert from VMChangeSet with non-materialized Aggregator V1 deltas to ChangeSet.",
+            ));
+        }
+        if !delayed_field_change_set.is_empty() {
+            return Err(code_invariant_error(
+                "Cannot convert from VMChangeSet with non-materialized Delayed Field changes to ChangeSet.",
+            ));
+        }
+
         let mut write_set_mut = WriteSetMut::default();
-        write_set_mut.extend(resource_write_set.into_iter().map(|(k, (v, _))| (k, v)));
+        write_set_mut.extend(
+            resource_write_set
+                .into_iter()
+                .map(|(k, v)| {
+                    Ok((
+                        k,
+                        v.try_into_concrete_write().ok_or_else(|| {
+                            code_invariant_error(
+                                "Cannot convert from VMChangeSet with non-materialized write set",
+                            )
+                        })?,
+                    ))
+                })
+                .collect::<Result<Vec<_>, _>>()?,
+        );
         write_set_mut.extend(module_write_set);
         write_set_mut.extend(aggregator_v1_write_set);
 
         let events = events.into_iter().map(|(e, _)| e).collect();
         let write_set = write_set_mut
             .freeze()
             .expect("Freezing a WriteSet does not fail.");
-        StorageChangeSet::new(write_set, events)
+        Ok(StorageChangeSet::new(write_set, events))
     }
 
-    /// Converts VM-native change set into its storage representation with fully
-    /// serialized changes. The conversion fails if:
-    /// - deltas are not materialized.
-    /// - resource group writes are not (combined &) converted to resource writes.
-    pub fn try_into_storage_change_set(self) -> anyhow::Result<StorageChangeSet, VMStatus> {
-        if !self.aggregator_v1_delta_set.is_empty() {
-            return Err(VMStatus::error(
-                StatusCode::DATA_FORMAT_ERROR,
-                err_msg(
-                    "Cannot convert from VMChangeSet with non-materialized Aggregator V1 deltas to ChangeSet.",
-                ),
-            ));
-        }
-        if !self.delayed_field_change_set.is_empty() {
-            return Err(VMStatus::error(
-                StatusCode::DATA_FORMAT_ERROR,
-                err_msg(
-                    "Cannot convert from VMChangeSet with non-materialized Delayed Field changes to ChangeSet.",
-                ),
-            ));
-        }
-        if !self.resource_group_write_set.is_empty() {
-            return Err(VMStatus::error(
-                StatusCode::DATA_FORMAT_ERROR,
-                err_msg(
-                    "Cannot convert from VMChangeSet with non-combined resource group changes.",
-                ),
-            ));
-        }
-        Ok(self.into_storage_change_set_unchecked())
+    pub fn concrete_write_set_iter(&self) -> impl Iterator<Item = (&StateKey, Option<&WriteOp>)> {
```

Comment:
I think I inspected all call-sites of this method and seems we can make it return &WriteOp instead of Option? Non-test uses are only in genesis, so it seems we might even be able to go further and extract and re-use the asserts for aggregator_v1 & delayed fields being empty with try_into_storage_change_set